### PR TITLE
Pfs 40 approve pending invoice adjustment

### PR DESCRIPTION
**Pull request by Nicholassully**

### CODE REVIEW

**Review by josephsmith0705 — DISMISSED** [review]

File: cypress/e2e/pending-invoice-adjustments.cy.js [hunk 1]
```diff
@@ -1,39 +1,44 @@
-describe("Pending Invoice Adjustments Tab", () => {
-    beforeEach(() => {
+// describe("Pending Invoice Adjustments Tab", () => {
+//     beforeEach(() => {
+//         cy.visit("/clients/1/pending-invoice-adjustments");
+//     });
```
Comment (denialanderror) [inline cypress/e2e/pending-invoice-adjustments.cy.js:4]: Is this meant to be here?

File: finance-api/cmd/api/server.go [hunk 2]
```diff
@@ -27,6 +28,7 @@ func (s *Server) SetupRoutes() {
 	http.HandleFunc("GET /clients/{id}/invoices", s.getInvoices)
 	http.HandleFunc("GET /clients/{id}/fee-reductions", s.getFeeReductions)
 	http.HandleFunc("GET /clients/{id}/invoice-adjustments", s.getInvoiceAdjustments)
+	http.HandleFunc("POST /clients/{id}/invoice-adjustments/{ledgerId}", s.updatePendingInvoiceAdjustment)
```
Comment (denialanderror) [inline finance-api/cmd/api/server.go:31]: If this is updating an existing resource, it should be a PUT, not a POST

File: finance-api/internal/store/queries/ledger_allocation.sql [hunk 3]
```diff
@@ -4,3 +4,33 @@ insert into ledger_allocation (id, ledger_id, invoice_id, datetime, amount, stat
...
+--              inner join invoice i on i.id = lc.invoice_id
+--              inner join finance_client fc on fc.id = i.finance_client_id
+--     where fc.client_id = $1 and l.id = $2
+--       and l.type IN ('CREDIT MEMO', 'CREDIT WRITE OFF')
+-- )
+-- UPDATE ledger_allocation
+-- SET status = 'APPROVED'
+-- FROM filtered_ledger_allocation fla
+-- WHERE ledger_allocation.id = fla.id
+-- returning ledger_allocation.*;
```
Comment (denialanderror) [inline finance-api/internal/store/queries/ledger_allocation.sql:22]: Commented out code

File: finance-api/internal/service/update_pending_invoice_adjustment.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	transaction := s.Store.WithTx(tx)
+	ledgerIdTransformed := int32(ledgerId)
+
+	_, err = transaction.UpdateLedgerAdjustment(ctx, ledgerIdTransformed)
+	if err != nil {
+		return err
+	}
+
+	_, err = transaction.UpdateLedgerAllocationAdjustment(ctx, ledgerIdTransformed)
```
Comment (denialanderror) [inline finance-api/internal/service/update_pending_invoice_adjustment.go:30]: If you are always throwing away the returned values, why do you return them?

File: finance-api/internal/store/queries/ledger_allocation.sql [hunk 5]
```diff
@@ -4,3 +4,33 @@ insert into ledger_allocation (id, ledger_id, invoice_id, datetime, amount, stat
...
+    from ledger l
+             inner join ledger_allocation lc on lc.ledger_id = l.id
+    where l.id = $1
+      and l.type IN ('CREDIT MEMO', 'CREDIT WRITE OFF')
+)
+UPDATE ledger_allocation
+SET status = 'APPROVED'
+FROM filtered_ledger_allocation fla
+WHERE ledger_allocation.id = fla.id
+returning ledger_allocation.*;
```
Comment (denialanderror) [inline finance-api/internal/store/queries/ledger_allocation.sql:36]: This is exactly the same query as the one in `ledger.sql`

File: finance-hub/internal/api/update_pending_invoice_adjustment.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"fmt"
+	"net/http"
+)
+
+func (c *ApiClient) UpdatePendingInvoiceAdjustment(ctx Context, clientId int, ledgerId int) error {
+
+	url := fmt.Sprintf("/clients/%d/invoice-adjustments/%d", clientId, ledgerId)
+	req, err := c.newBackendRequest(ctx, http.MethodPost, url, nil)
```
Comment (denialanderror) [inline finance-hub/internal/api/update_pending_invoice_adjustment.go:11]: Should be a PUT

File: finance-api/cmd/api/update_pending_invoice_adjustment.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	ledgerId, _ := strconv.Atoi(r.PathValue("ledgerId"))
+	err := s.Service.UpdatePendingInvoiceAdjustment(ledgerId)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```
Comment (denialanderror) [inline finance-api/cmd/api/update_pending_invoice_adjustment.go:19]: We are not returning any data here, so the content-type is not needed and it should really return a 204 if we are being _reaaaaaally_ picky

File: finance-hub/internal/server/pending_invoice_adjustments.go [hunk 8]
```diff
@@ -35,23 +37,24 @@ func (h *PendingInvoiceAdjustmentsHandler) render(v AppVars, w http.ResponseWrit
...
 		out = append(out, PendingInvoiceAdjustment{
-			Id:               adjustment.Id,
+			Id:               strconv.Itoa(adjustment.Id),
 			Invoice:          adjustment.InvoiceRef,
 			DateRaised:       adjustment.RaisedDate,
 			AdjustmentType:   h.transformType(adjustment.AdjustmentType),
 			AdjustmentAmount: util.IntToDecimalString(adjustment.Amount),
 			Notes:            adjustment.Notes,
 			Status:           h.transformStatus(adjustment.Status),
+			ClientId:         clientId,
```
Comment (denialanderror) [inline finance-hub/internal/server/pending_invoice_adjustments.go:57]: Is it necessary to add the client ID to every adjustment? Couldn't this just be on the top level vars and accessed from there?

**Review by denialanderror — CHANGES_REQUESTED** [review]

File: finance-hub/internal/server/pending_invoice_adjustments.go (hunk 8 above)
Comment (Nicholassully) [reply finance-hub/internal/server/pending_invoice_adjustments.go:57]: Moved it out

File: finance-api/cmd/api/update_pending_invoice_adjustment.go (hunk 7 above)
Comment (Nicholassully) [reply finance-api/cmd/api/update_pending_invoice_adjustment.go:19]: Ok changed

File: finance-hub/internal/api/update_pending_invoice_adjustment.go (hunk 6 above)
Comment (Nicholassully) [reply finance-hub/internal/api/update_pending_invoice_adjustment.go:11]: done

File: finance-api/internal/store/queries/ledger_allocation.sql (hunk 5 above)
Comment (Nicholassully) [reply finance-api/internal/store/queries/ledger_allocation.sql:36]: It was and have changed it

File: finance-api/internal/service/update_pending_invoice_adjustment.go (hunk 4 above)
Comment (Nicholassully) [reply finance-api/internal/service/update_pending_invoice_adjustment.go:30]: Good point changed

File: finance-api/cmd/api/server.go (hunk 2 above)
Comment (Nicholassully) [reply finance-api/cmd/api/server.go:31]: done

File: cypress/e2e/pending-invoice-adjustments.cy.js (hunk 1 above)
Comment (Nicholassully) [reply cypress/e2e/pending-invoice-adjustments.cy.js:4]: no and gone

**Review by denialanderror — APPROVED** [review]

**Pull request merged**